### ISSUE TITLE
[Discussion] 1.6.0 Roadmap

### ISSUE THREAD

**Issue by szha** [apache/incubator-mxnet#15589]
Let's start a discussion here about the roadmap towards 1.6.0. We are looking for:

New features that are useful to your research and development.
Improvements and patches to existing features.
If you have any item that you'd like to propose to have in the roadmap, please do:

Create (or locate existing) issue/pull request for the item, note the issue/pull request number.
Comment in this issue: 1) the above issue number, 2) one sentence of what the item is about and why it's useful to you.
Indicate whether you'd be willing to help out on the item.
Share the ETA if you're driving the item and have an guesstimate on when it will be done.

Feel free to include items that weren't included in past roadmap discussions that you still wish to include in this release.
cc @apache/mxnet-committers

**Comment by sandeep-krishnamurthy** [comment]
1. Can we work towards INT-64 support (Large Tensor) as DEFAULT. This would make it easy for Large Tensor use cases such as DGL, recommendation system models to use MXNet easily. Today, they need to compile from source. (CC @apeforest @access2rohit)

**Comment by anirudhacharya** [comment]
> 1. Can we work towards INT-64 support (Large Tensor) as DEFAULT. This would make it easy for Large Tensor use cases such as DGL, recommendation system models to use MXNet easily. Today, they need to compile from source. (CC @apeforest @access2rohit)

There are a few regressions that have been introduced since the large tensor support was added. For example, there are a few hierarchical attention networks which were training fine until 1.3.1 but since the addition of large tensor support, results in NaNs in weight vectors and gradient calculations. We probably should ensure all regressions are fixed with regard to this PR. 

Related issues and PRs that might be relevant - 
 - https://github.com/apache/incubator-mxnet/pull/11742#issuecomment-449181485
 - https://github.com/apache/incubator-mxnet/issues/14496
 - https://github.com/apache/incubator-mxnet/pull/14570

**Comment by pengzhao-intel** [comment]
@sandeep-krishnamurthy Agree INT64 enhancements.
We plan to upgrade MKLDNN to 1.0 in r1.6 so that MKLDNN backend can work for INT64 index.

**Comment by pengzhao-intel** [comment]
CPU related proposal for 1.6 (keep updating in the following several days).

1. MKL-DNN upgrade to 1.0
-  the proposal will be sent out soon in dev@   @TaoLv 

2. Quantization API improvement @xinyu-intel

**Comment by wkcn** [comment]
Agree INT64 enhancements, too.
For the regression problem, a good solution is to choose data type by the data size in the performance bottleneck.

**Comment by Neutron3529** [comment]
I think MXNet needs a relocatable .dll/.so file
`libmxnet.dll` is too large and takes too much time to load in my Windows 10

I asked how to decrease the dll size (since there is too much `arch`s in the single .dll file and what I want is just `-arch=sm_61` for my GTX 1060)

The reply is to use `nvprune`, I tried it and it gives me an error:
```
nvprune fatal   : Input file 'libmxnet.dll' not relocatable
```
Make the `libmxnet.dll`/`libmxnet.so` relocatable will provides the possibility to further decrease the size of the `dll` file. And may further more decrease the import time.

**Comment by sandeep-krishnamurthy** [comment]
It would be nice to add support for default checkpointing in Estimator API.
Also, can we plan to graduate Estimator API in MXNet 1.6? This is super useful API for MXNet users.

@roywei - Any comments?

**Comment by ptrendx** [comment]
We have multiple improvements to BERT inference and training speed that we would like to be part of 1.6 release:
[x] Softmax optimizations (#15545 )
[ ] Pointwise fusion for GPU (#15167 )
[ ] Eliminate common expressions (#15657 )
[ ] Bias speed improvements (#16039 )
[ ] Aggregated AdamW optimizer (not yet PR'ed)
[ ] Aggregated zeroing of the gradients (not yet PR'ed)
[ ] Aggregated sum of squares operator (also used in LARS, @Caenorst is working on a PR)
[ ] Embedding gradient optimization (not yet PR'ed)
[ ] Faster multihead attention operator (not yet PR'ed)

**Comment by roywei** [comment]
Moving fixing nightly failure from 1.5.1 scope to 1.6.0 as they are failing on master branch not 1.5.x branch.
https://github.com/apache/incubator-mxnet/issues/15613#issuecomment-516937546

> 
> nightly test failure need to be fixed:
> #15374
> 
> http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/NightlyTestsForBinaries/detail/master/395/pipeline/

**Comment by szha** [comment]
@reminisce @haojin2 given that numpy operators would be a major topic in 1.6 release, it would be great if you could add what you intend to include in this release.

**Comment by hkingtswcbyy** [comment]
If I want to use mxnet.gluon.nn.Conv2D to get a depthwise conv layer, I need to explicit set the group argument. Can this be infered automatically?

**Comment by reminisce** [comment]
As for NumPy compatibility, we would like to add
1. A set of NumPy operators mostly with CPU/GPU forward/backward supports.
2. A new `ndarray` class that replicates NumPy's `ndarray` in most ways. (Differences will be documented).
3. NumPy op integration with Gluon components, including layers, model zoo, parameters, data loaders, etc.

**Comment by KellenSunderland** [comment]
TRT is now working from the CPP package.  I think to consider it a released feature we'd want to update documentation and possible target the new TRT version (6).

**Comment by anirudh2290** [comment]
I am working on an interface for multi threaded inference in MXNet and it would be great if it could go in 1.6.

**Comment by szha** [comment]
@anirudh2290 this sounds like a larger change. Would you link to the RFC for it?

**Comment by anirudh2290** [comment]
@szha yes I am planning to add a RFC this week.

**Comment by leezu** [comment]
For reference, users are confused that Large Tensor Support was enabled in MXNet 1.4 and then disabled again in 1.5. Reference: https://github.com/apache/incubator-mxnet/pull/13418
For 1.6, can we improve the error message to suggest people to compile with Large Tensor Support?

**Comment by access2rohit** [comment]
@leezu I am working on it. Here is the WIP PR: https://github.com/apache/incubator-mxnet/pull/16570

**Comment by jonatan1626** [comment]
Hi I was doing some testing for mxnet 1.6.x and 1.5.1 and I noticed some performance issues in training you can find more details here: [16845](https://github.com/apache/incubator-mxnet/issues/16845)
